cpp/src/strings/char_types/char_types.cu

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/char_types/char_types.hpp>
+#include "../utilities.hpp"
+#include "../utilities.cuh"
+
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+
+//
+//
+std::unique_ptr<cudf::column> is_characters_of_type( strings_column_view strings,
+                                                     string_character_types types,
+                                                     rmm::mr::device_memory_resource* mr,
+                                                     cudaStream_t stream = 0)
+{
+    auto strings_count = strings.size();
+    auto execpol = rmm::exec_policy(0);
+    auto strings_column = column_device_view::create(strings.parent(),stream);
+    auto d_column = *strings_column;
+
+    rmm::device_buffer null_mask;
+    cudf::size_type null_count = d_column.null_count();
+    if( d_column.nullable() ) // copy null_mask
+        null_mask = rmm::device_buffer( d_column.null_mask(),
+                                        gdf_valid_allocation_size(strings_count),
+                                        stream, mr);
+
+    // create output column
+    auto results = std::make_unique<cudf::column>( cudf::data_type{cudf::INT8}, strings_count,
+        rmm::device_buffer(strings_count * sizeof(int8_t), stream, mr),
+        null_mask, null_count);
+    auto results_view = results->mutable_view();
+    auto d_results = results_view.data<int8_t>();
+    //
+    auto d_flags = get_character_flags_table();
+    // set the output values by checking the character types
+    thrust::for_each_n(execpol->on(stream),
```

Comment:
Would get more parallelism by parallelizing over all characters in the data array, using a segmented reduction. Would need to generate a key array where all characters of a given string have index of the string.

Reply:
Only if all the characters were the same width and their corresponding case character matched their width. The problem is trying to place an individual character in the correct spot in the output. This would require creating 32-bit or 64-bit offset array for every character.